src/rpc/src/chainStorageWatcher.ts

```diff
@@ -0,0 +1,167 @@
+  let nextQueryTimeout: number | null = null;
+
+  const queueNextQuery = () => {
+    if (isQueryInProgress || !watchedPathsToSubscribers.size) {
+      return;
+    }
+
+    if (isNewPathWatched) {
+      // If there's new paths to watch, schedule another query very soon.
+      if (nextQueryTimeout) {
+        window.clearTimeout(nextQueryTimeout);
+      }
+      nextQueryTimeout = window.setTimeout(queryUpdates, newPathQueryDelayMs);
+    } else {
+      // Otherwise, refresh after a normal interval.
+      nextQueryTimeout = window.setTimeout(
+        queryUpdates,
+        randomRefreshPeriod(refreshLowerBoundMs, refreshUpperBoundMs),
+      );
+    }
+  };
+
+  const queryUpdates = async () => {
+    isQueryInProgress = true;
+    nextQueryTimeout = null;
+    isNewPathWatched = false;
+
+    const paths = [...watchedPathsToSubscribers.keys()].map(pathKey =>
+      JSON.parse(pathKey),
+    );
+
+    if (!paths.length) {
+      isQueryInProgress = false;
+      return;
+    }
+
+    try {
+      const data = await batchVstorageQuery(rpcAddr, unserialize, paths);
+      watchedPathsToSubscribers.forEach((subscribers, path) => {
+        const { blockHeight, value } = data[path];
+        const lastValue = latestValueCache.get(path);
+
+        if (
+          lastValue &&
+          (blockHeight === lastValue[0] ||
+            JSON.stringify(value) === lastValue[0])
+        ) {
+          // The value isn't new, don't emit.
+          return;
+        }
+
+        latestValueCache.set(path, [
+          blockHeight ?? JSON.stringify(value),
+          value,
+        ]);
+
+        subscribers.forEach(s => {
+          s.onUpdate(harden(value));
+        });
+      });
+    } finally {
+      isQueryInProgress = false;
+      queueNextQuery();
+    }
+  };
+
+  const stopWatching = (pathKey: string, subscriber: Subscriber<unknown>) => {
+    const subscribersForPath = watchedPathsToSubscribers.get(pathKey);
+    if (!subscribersForPath) {
+      throw new Error('Trying to unsubscribed from unwatched path');
+    }
+
+    if (subscribersForPath.size === 1) {
+      watchedPathsToSubscribers.delete(pathKey);
+      latestValueCache.delete(pathKey);
+    } else {
+      subscribersForPath.delete(subscriber);
+    }
+  };
+
+  const queueNewPathForQuery = () => {
+    if (!isNewPathWatched) {
+      isNewPathWatched = true;
+      queueNextQuery();
+    }
+  };
+
+  const watchLatest = <T>(
+    path: [AgoricChainStoragePathKind, string],
+    onUpdate: (latestValue: T) => void,
+  ) => {
+    const pathKey = JSON.stringify(path);
+    const subscriber = makePathSubscriber(onUpdate);
+
+    const latestValue = latestValueCache.get(pathKey);
+    if (latestValue) {
+      (subscriber as Subscriber<unknown>).onUpdate(harden(latestValue[1]));
+    }
+
+    const samePathSubscribers = watchedPathsToSubscribers.get(pathKey);
+    if (samePathSubscribers !== undefined) {
+      samePathSubscribers.add(subscriber as Subscriber<unknown>);
```

Comment:
these three casts to unknown would go away if the map were instead a fully typed object